### ISSUE TITLE
Batching shard requests

### ISSUE THREAD

**Issue by Totktonada** [tarantool/graphql.0#98]
## Breadth-first traversal for core/executor.lua

Don't call evaluateSelections from completeValue, but call collectFields on
each object, remove fields that are present. Save parameters for getFieldEntry
call (they forms open requests lists, one for each field).

Get requests list and resolve it: get list of shard requests, resolve it in
batch (get objects or lists of objects), invoke postprocessing function, call
getFieldEntry to form next open requests list.

## Batch requests to shard

Call to accessor:select returns parameters for shard request and 
postprocessing function. TBD: limit.

**Comment by Totktonada** [comment]
I look into the proposal above deeper and I have to say two points:

1. It involves even more captures than we have now that lead to unclear link btw abstractions we have, and just hard to understand.
2. Don't sure it will work, because… ugh, I just befuddled with objects / capture to resolve objects that were proposed to use.

So, below I proposed another view of an executor and related problems.

## Execution program / query plan

Now we perform depth-first traversal and strictly successive (blocking)
network requests for each connected / nested field. That is extremely slow.

Imagine we have implemented generator of execution programs by queries 
(working with breadth-first traversal) and an executor (interpreter) for such
programs. I'll show example of a query and the generated program and will
deep a bit into problems we'll have with trying to execute that program in an
optimal way.

### Create execution program

Input: enhanced graphql types (or graphql types + extra info) + query ast.
Output: execution program.

### Motivating example

Example query:

```
query get_order($order_id: String) {
    order(order_id: $order_id) {
        description
        user { // 1:1 connection (o.user_id = u.user_id)
            first_name
            last_name
        }
    }
}
```

The program should looks like:

```
// define query variables
VARIABLES {
    $order_id,
}

// fetch orders
// ------------

// fetch list of tuples and produce objects; note: we have an one tuple here,
// but the type is a list of object, so we use SELECT and don't use GET
$tuples_1 = SELECT {
    collection = "order",
    index = "order_id",
    key = $order_id,
}
$objects_1 = FOREACH $tuple in $tuples_1 {
    YIELD POSTPROCESS {
        collection = "order",
        tuple = $tuple,
    }
}
$selections_1 = FOREACH $object in $objects_1 {
    YIELD OBJECT {
        description = $object.description,
    }
}

// fetch users
// -----------

// fetch list of tuples using fields of parent objects
$tuples_2 = FOREACH $object in $objects_1 {
    // 1:1 connection, so GET is used
    YIELD GET {
        collection = "user",
        index = "user_id",
        key = $object.user_id,
    }
}

// produce objects from fetched tuples
FOREACH $tuple in $tuples_2 COUNTER $i {
    $object_2 = POSTPROCESS {
        collection = "order",
        tuple = $tuple,
    }
    $selection_2 = OBJECT {
        first_name = $object_2.first_name,
        last_name = $object_2.last_name,
    }
    $selections_1[i].user = $selection_2
}

RETURN $selections_1
```

XXX: unify objects / selections creation to give better idea how to generate
such program.

### Language

Operators:

* SELECT (indexed or performing fullscan)
* GET (indexed)
* POSTPROCESS (tuple -> object)
* OBJECT: create an object with specified set of fields
* FOREACH + YIELD
* RETURN
* assignment to a new variable (=)
* assignment to a new field of an object (=)

Operations:

* get / set field of an object
* get item of a list

### Problems

The program above have the following problems (Nth problem assumes that
(N-1)th was resolved).

1. We perform GETs from the "user" collection (they are requests over network)
   one-by-one, but we need to aggregate GETs / SELECTs into batches / blocks
   to minimize network traffic and overhead of network-related machinery.

2. We perform such GETs (or GET blocks) successive, but we need to have many
   network requests in-the-fly to better utilize network throughput.

3. We perform breadth-first traversal, but that can lead to network stalls (we
   don't prepare / put to the queue ready requests for connected / nested
   fields) or memory overflow (in case of frontend-side filtering and large
   SELECT on the top level); we need to have several instruction pointers (and
   some kind of dependency btw IPs) / several connected programs.

### TBD

Elaborate connected programs approach.

**Comment by Totktonada** [comment]
### Notes on Hyracks

Hyracks executes jobs once it ready (it manages job dependecies).

Hyracks and Hadoop MapReduce partions/maps each intermediate result list by a
key and send to corresponding nodes.

Attaching field value is the map operation by their nature.

Map outputs a list for each input item is okay.

Things to consider in the future:
* push-down tasks to nodes more or less as is (not just shard requests, but a
  program for execute by a storage-local interpreter);
* now we generating objects to match it with filters, but we can convert
  filters into 'tuple fields' terms; can be profitable.

Hyracks task in our terms: receive list of tuple (block), push resulting
blocks.

Hyracks has N:M connectors where N is a count of input producers and M is a
count of results consumers; a connector performs repartitioning of a task
results.

* How it looks for frondend executor + storage executors (no full mesh)?
* How it looks for single frontend executor (no storage executors)?

I guess that root tasks in Hyracks are scheduled to nodes contain its initial
data. The frontend-executor (with/without storage executors) case can looks
similar, but we need to add extra data push tasks to storages and make it
pushing its output to actual data processor on frontend / related backend
node.

Hyracks sends a block when we produce enough objects to form a block.

Hyracks frame (block) is about 32 KB (at least for an one of presented
workloads).

Hyracks works beyod map-reduce, that faster in some cases. Its hash join
receives two inputs w/o full sort as I understand.

**Comment by Totktonada** [comment]
### Pipelined tasks

Let's try to rewrite above program using pipelined tasks.

```
// name: fetch orders
// input: $order_id
// output: order tuples
// note: each invoke produces list of task to push further
SELECT {
    collection = "order",
    index = "order_id",
    key = $order_id,
} 
```
  
```
// name: postprocess orders
// input: $order_tuple
// output: order object
POSTPROCESS {
    collection = "order",
    tuple = $order_tuple,
}
```                    
  
```                  
// name: select order fields
// input: $order_object
// output: cut order object
OBJECT {               
    description = $order_object.description,
} 
```

```
// name: gen user fetch ids
// input: $order_object
// output: user_id
RETURN $order_object.user_id
```

```
// name: fetch users
// input: $user_id
// output: user object
GET {
    collection = "user",
    index = "user_id",
    key = $user_id,
}
```

...skipped user-related tasks that almost same as for orders...

```
// name: attach users to orders
// input: $cut_order_object, $cut_user_object
RETURN $cut_order_object + {user = $cut_user_object}
```

Tasks expressed in form of processing an one object, but under hood the
transfers performed by blocks.

```
gen_order_fetch_id
  |
  v
fetch_orders (storages)
  |
  v
merger
  |
  v
postprocess_orders -+
  |                  \
  v                   v
select_order_fields  gen_user_fetch_ids
  |                    |
  |                    v
  |                  fetch_users (storages)
  |                    |
  |                    v
  |                  merger
  |                    |
  |                    v
  |                  postprocess_users
  |                    |
  |                    v
  |                  select_user_fields
  |                   /
  v                  v
attach_users_to_orders
```

### Implementation questions

1. How to manage pipelined execution, which tarantool entities need to be
   involved? Fiber + fiber channel for local pipes and queue + netbox for
   over-machine pipes?
2. Should blocks be written explicitly or just use yield to allow a consumer
   read the input side of a local pipe?

### PoC constraints

1. Storage nodes supports only data fetch tasks.
2. No storage-storage connections.

We omit partioning phase here, because virtually all execution performing by a
frontend machine.

### Problems

1. In that simple example we attach Nth user to Nth order, but 1:N connections
   or third nesting level require more complex approaches. The problem is that
   we: a) lack of structure when flatten sequence of SELECTs results, b) lack
   of parent address when return objects.

We can flatten the response tree into a list, but always have some kind of
address within each tuple/object. Then we can reconstruct the tree from that
list. The address can be (item:N, item:M, field:field_name), it more or less
similar to avro-schema namespace.

### Notes

Push-down filtering is better to make on tuples with appropriate filter
converting, because:

1. It should be faster than unpack (Hyracks does the same).
2. Merger works with tuples.

TBD: will we hide resident node of a task somehow or make it part of a query
plan?

TBD: evolve the idea with flat tree representation and reconstruction
afterwards.

TBD: write query plan for examples with 1:N connection, a filter, three
nesting level, a filter over a connection, a directive, an union connection.

TBD: consider scheme with pushing from one storage node to an another.

**Comment by Totktonada** [comment]
## Near plans / PoC

1. split resolve to the two phases:
   - preparing a shard request;
   - executing a shard request + converting the tuple into the object;
2. implement breadth-first executor on top of query ast + schema;
3. inject select-ids remote procedure to storages;
4. use merger separatelly from mr_select to merge select-ids results;
5. merge sequence of same-typed requests;
6. productize branch: writing ldocs, tests and so on.

Maybe it worth to merge needed schema and query information into an one
structure to achieve the following goals:
                                         
* get rid of graphql-lua data structures (tangled a bit, IMO);
* have an one tree to traversal, not several trees for parallel traverse;                                                                                                     
* make basis for future query plan.

**Comment by Totktonada** [comment]
The PoC was implemented in PR #113.
